Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -4,34 +4,231 @@
+        insert(type, key, values[i], errcode);
+   }
+   return errcode;
+}
+
+// searchHashTable takes in an integer key and a type
+// calls the helper function to hash the received key to correspond to an expected index within the hash table
+// if the key at the position in the test hashTable indicated by the hashing function matches the received key then return a SUCCESS code
+// if a mistmatch occurs return an ERROR code
+int searchHashTable(int key, int * success) {
+    if (key > SIZE || key < 0) return ERROR;
+    int hashIndex = hashing(key);
+    if (testHash[hashIndex]->key == key) {
+        *success = SUCCESS;
+    }
+}
+
+
+// testSearch
+// First creates a testHash with the testing key/value pairs. 
+// calls the helper function searchHashTable on the testHash
+// if the key/value pair is found successfully and the expectedOutcome matches a SUCCESS code 0 is returned then a success message is printed and a SUCCESS code is returned
+// if the value is not found then a failure message is printed and an ERROR -1 is returned
+int testSearch(int key, int expectedOutcome) {
+    //build the testHash
+    int keys[10] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+    int value[10]= { 10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
+    buildHashTable( 9, keys, value);
+    int success = ERROR;
+    searchHashTable(key, &success);
+    if (success == SUCCESS && expectedOutcome == SUCCESS) {
+        printf("The key %d was successfully found as expected. \n", key);
+        return SUCCESS;
+    } else if (success == ERROR && expectedOutcome == ERROR) {
+        printf("The key %d was not found as expected. \n", key);
+        return SUCCESS;
+    }
+    printf("Mismatch: An unknown error has occurred. \n");
+    return ERROR;
 }
 
+// buildSearchTests
+// creates 5 test cases expected to pass the testSearch function
+// creates 5 test cases (out of range) expected to fail the testSearch function
+void buildSearchTests() {
+printf("Build Search Test Passes: \n");
+    testSearch(1, SUCCESS);
+    testSearch(8, SUCCESS);
+    testSearch(3, SUCCESS);
+    testSearch(4, SUCCESS);
+    testSearch(9, SUCCESS);
+
+    //5 Fail
+    printf("Build Search Test Fails: \n");
+    testSearch(99, ERROR);
+    testSearch(-9, ERROR);
+    testSearch(11, ERROR);
+    testSearch(20, ERROR);
+    testSearch(-20, ERROR);
+}
 
+// deleteItem takes in a key to delete and an int to signify success of deletion
+// finds the key of the provided Item then passes it to the helper function to hash it and find the corresponding index
+// if the key at the position in the hashTable indicated by the hashing function matches the received key then sets an empty delete Item equal to the Item at the hashIndex
+// if a failure occurs then the success code remains as ERROR
+void deleteItem(int key, int * success) {
+    int hashIndex = hashing(key);
+    if (testHash[hashIndex]->key == key) {
+        free(testHash[hashIndex]);
+        insert(9, key, NULL, *success);
```

Comment:
`9` is a magic number.  I imagine this will go away with the global state purge.

Reply:
Updated in [commit a0219c38606fafc63d5c641e0623442f105ff2b](https://github.com/kcarrel/Training-Notes/pull/12/commits/a0219c38606fafc63d5c641e0623442f105ff2b7) to move away from global state/magic numbers.